pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,190 @@
+// It periodically check for Node events (creating/deleting) to apply appropriate
+// tags to resources.
+type TaggingController struct {
+	controllerOptions options.TaggingControllerOptions
+	kubeClient        clientset.Interface
+	nodeLister        v1lister.NodeLister
+	cloud             *awsv1.Cloud
+
+	// Value controlling TaggingController monitoring period, i.e. how often does TaggingController
+	// check node list. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+
+	// A map presenting the node and whether it currently exists
+	currNodes map[string]bool
+
+	// A map representing nodes that were ever part of the cluster
+	totalNodes map[string]*v1.Node
+
+	// Representing the user input for tags
+	tags map[string]string
+
+	// Representing the resources to tag
+	resources []string
+}
+
+// NewTaggingController creates a NewTaggingController object
+func NewTaggingController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration,
+	tags map[string]string) (*TaggingController, error) {
+
+	awsCloud, ok := cloud.(*awsv1.Cloud)
+	if !ok {
+		err := fmt.Errorf("tagging controller does not support %v provider", cloud.ProviderName())
+		return nil, err
+	}
+
+	tc := &TaggingController{
+		kubeClient:        kubeClient,
+		nodeLister:        nodeInformer.Lister(),
+		cloud:             awsCloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		currNodes:         make(map[string]bool),
+		totalNodes:        make(map[string]*v1.Node),
+		tags:              tags,
+	}
+	return tc, nil
+}
+
+// Run will start the controller to tag resources attached to the cluster
+// and untag resources detached from the cluster.
+func (tc *TaggingController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	wait.UntilWithContext(ctx, tc.MonitorNodes, tc.nodeMonitorPeriod)
+}
+
+func (tc *TaggingController) MonitorNodes(ctx context.Context) {
+	nodes, err := tc.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Fatalf("error listing nodes: %s", err)
+	}
+
+	for k := range tc.currNodes {
+		tc.currNodes[k] = false
+	}
+
+	var nodesToTag []*v1.Node
+	for _, node := range nodes {
+		if _, ok := tc.currNodes[node.GetName()]; !ok {
+			nodesToTag = append(nodesToTag, node)
+		}
+
+		tc.totalNodes[node.GetName()] = node
+		tc.currNodes[node.GetName()] = true
+	}
+	tc.tagNodesResources(nodesToTag)
+
+	var nodesToUntag []*v1.Node
+	for nodeName, existed := range tc.currNodes {
+		if existed == false {
+			nodesToUntag = append(nodesToUntag, tc.totalNodes[nodeName])
+		}
+	}
+	tc.untagNodeResources(nodesToUntag)
+}
+
+// tagNodesResources tag node resources from a list of nodes
+// If we want to tag more resources, modify this function appropriately
+func (tc *TaggingController) tagNodesResources(nodes []*v1.Node) {
+	for _, node := range nodes {
+		nodeTagged := false
```

Comment:
Seems redundant?

Reply:
Could you elaborate? I'd like to keep it here cause it'll make the code easier to read and I want to keep track if resource is being tagged correctly or not.

Reply:
I was referring to the statement `nodeTagged := false`. You're initializing it in the next line anyway so this line seems redundant.

Reply:
Ahh, I see what you meant. Perhaps it would be clearer in the next revision, here's a sneak-peek
```
		nodeTagged := false

		for _, resource := range tc.resources {
			if resource == opt.Instance {
				nodeTagged = tc.tagEc2Instances(node)
			}
		}
```